### ISSUE TITLE
Choose to not expose version number

### ISSUE THREAD

**Issue by rbricheno** [mastodon/mastodon#21925]
### Pitch

I'd rather not advertise the version number of Mastodon that I'm running. It would be convenient to be able to disable this through the Admin control panel.

### Motivation

Displaying version number by default makes it very easy to search for sites running a version of Mastodon with known security bugs (for example).

**Comment by LeoEurope** [comment]
Knowing the version number can be quite useful for people, though. I don't think I'd sign up on a server that is still running 3.0 for instance.

How about displaying only the major version number? So 3.5 but not 3.5.5 and 4.0 but not 4.0.2. That way attackers can't see if you have yet to patch the latest fixes. You could decide to only show the patch level to logged in users.

**Comment by rbricheno** [comment]
@LeoEurope in my case that's a moot point as no users may sign up to my instance, it's purely personal. But sure, no harm in this being a selectable option for site admins:

- Do not display version number
- Display major version only
- Additionally, display full version number to logged in users

Selfishly, my problem is solved if I just have an easy way to hide the version number on my personal instance, which persists between upgrades, and doesn't require me to hack version.rb every time

**Comment by mc776** [comment]
If I'm shopping for instances, or corresponding with someone and wanting to know if an emoji reaction would show up, or trying to understand why a person is posting in some offputting way (weird abbreviations, #lots of #inline #hashtags, etc.), or trying to help some random stranger with some weird glitch(!) they're having, this information is helpful.

Knowingly running an old version of something that has a security hole sounds like a terrible idea anyway.

**Comment by connyduck** [comment]
Client apps need to see the version for feature detection.

**Comment by rbricheno** [comment]
If I wanted to support people who are shopping for instances, then I would presumably choose not to hide my version number. I'm just asking for the option to choose not to (on an instance-by-instance basis) where those use cases are not important.

@mc776 yes knowingly running an old version of software is often not a good idea. But advertising that you are running an old version of buggy software to the internet is also bad. To put it bluntly, disclosing the version of Mastodon running on a server, by default, presents a security risk. Hiding it contributes to a "defense-in-depth" strategy. Fortunately, this doesn't have to be a dichotomy, and I can update my Mastodon instance often while still not wanting to disclose what version I'm running.

@connyduck is this a specific client app, or does it affect the web browser for example? I'd have thought Mastodon<->Plemora interaction (for example) would not work if this were really a hard requirement for all client interactions. If it's only certain client apps that require this, I'd like to be able to make the informed choice about whether to support them, or to not disclose my version information. As I alluded to earlier, I am currently running with a manually edited version.rb which provides a fake version number, and nothing appears to have broken for my use case.

**Comment by LeoEurope** [comment]
> Client apps need to see the version for feature detection.

Presumably these clients only need the API to confirm that, they don't have to scrape it off the website.

**Comment by pedroCX486** [comment]
> > Client apps need to see the version for feature detection.
> 
> Presumably these clients only need the API to confirm that, they don't have to scrape it off the website.

Anyone looking to exploit your site would then check with your API the version running. Or they would simply try to run the exploit anyway regardless. This “feature” is moot.

**Comment by rbricheno** [comment]
@pedroCX486 you're assuming that attacks are targeted rather than opportunistic. Trivial example: google "View source code · v4.0.1"

**Comment by pedroCX486** [comment]
> @pedroCX486 you're assuming that attacks are targeted rather than opportunistic. Trivial example: google "View source code · v4.0.1"

I never said or assumed that, most attacks are made using crawlers and they would also check the API if the info was hidden from the front-end.

**Comment by mc776** [comment]
> I don't know if other instances may need to know the version number of your instance for federation/communication purposes: so that could be a problem.

This does bring up another issue - if you're running an instance that you know has a security hole so big you need to keep people from knowing about it, I would think that's something someone who is sending their users' posts (and DMs) to your server would *really* want to know.

**Comment by rbricheno** [comment]
Without wanting to send this specific feature request massively off-topic, I've been thinking quite a bit about this and wrote up my thoughts here:

https://rwhb.me/20221204/

I'd be (somewhat) interested in discussing it more (including being told why I'm wrong!) in a setting that isn't specifically about this small feature request (e.g. feel free to contact me on Mastodon).

**Comment by ThisIsMissEm** [comment]
I don't think this would actually help with protecting the security of an out-of-date instance, because why would an attacker care what version you're running? They're just going to try all known attacks hoping something will work (much like why non-wordpress sites almost always have error logs of 404s where an attack is trying to attack a domain based on a known wordpress vulnerability.

Hiding the version of Mastodon that you're running will not save you from known vulnerabilities on your specific version. The best way to stay safe is to upgrade as new versions become available.

(also of note is that Mastodon has a responsible disclosure policy, and [reports of security issues](https://github.com/mastodon/mastodon/blob/main/SECURITY.md) are embargoed until a fix is released to the public)
